### PR TITLE
Add tags to support regular expressions without breaking the old ones in yaml conf

### PR DESCRIPTION
**Pull request by zhangguanzhang**

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM
But tests are failing.
@mtrmac @vrothberg PTAL

**Comment by zhangguanzhang** [comment]
tks,I will try this

**Comment by zhangguanzhang** [comment]
there is some err in travis-ci, Let me check the code

**Comment by zhangguanzhang** [comment]
I changed the code, and now the travis-ci test has passed

**Comment by rhatdan** [comment]
Thanks @zhangguanzhang

File: cmd/skopeo/sync.go [hunk 1]
```diff
@@ -278,21 +279,82 @@ func imagesToCopyFromRegistry(registryName string, cfg registrySyncConfig, sourc
...
 			if err != nil {
 				logrus.WithFields(logrus.Fields{
-					"tag": source,
-				}).Error("Error processing tag, skipping")
-				logrus.Errorf("Error getting image reference: %s", err)
-				continue
+					"repo":     imageName,
+					"registry": registryName,
+				}).Error("Error processing repo, skipping")
+				logrus.Error(err)
```
Comment (mtrmac) [inline cmd/skopeo/sync.go:355]: Missing `continue`, (incorrect error message?).

File: cmd/skopeo/sync.go [hunk 2]
```diff
@@ -317,6 +379,25 @@ func imagesToCopyFromRegistry(registryName string, cfg registrySyncConfig, sourc
...
 				continue
 			}
+
+			logrus.WithFields(logrus.Fields{
+				"repo":     imageName,
+				"registry": registryName,
+			}).Infof("Start filtering using the regular expression: %v", tags.(string))
+			for _, sReference := range allSourceReferences {
+				// get the tag names to match, [1] default is "latest" by .DockerReference().String()
+				if tagReg.Match([]byte(strings.Split(sReference.DockerReference().String(), ":")[1])) {
```
Comment (mtrmac) [inline cmd/skopeo/sync.go:389]: This should use the reference API for splitting tags.

Comment (mtrmac) [inline cmd/skopeo/sync.go:389]: Why not `MatchString`?

File: cmd/skopeo/sync.go [hunk 3]
```diff
@@ -278,21 +279,82 @@ func imagesToCopyFromRegistry(registryName string, cfg registrySyncConfig, sourc
...
-			source := fmt.Sprintf("%s:%s", repoName, tag)
 
-			imageRef, err := docker.ParseReference(source)
+		switch tags.(type) {
+		case []string, []interface{}, nil:
+			tagList := make([]string, 0)
+			if tagIns, ok := tags.([]interface{}); ok {
+				for _, tagValue := range tagIns {
+					switch tagValue.(type) {
+					case string, int, float64:
```
Comment (mtrmac) [inline cmd/skopeo/sync.go:289]: The `int`/`float64` types in here are scary. Are we really doing tag matches on a roundtrip between string/binary float/string?

File: cmd/skopeo/sync.go [hunk 4]
```diff
@@ -308,7 +370,7 @@ func imagesToCopyFromRegistry(registryName string, cfg registrySyncConfig, sourc
 				continue
 			}
 
-			sourceReferences, err = imagesToCopyFromRepo(imageRef, repoName, serverCtx)
+			allSourceReferences, err := imagesToCopyFromRepo(imageRef, repoName, serverCtx)
```
Comment (mtrmac) [inline cmd/skopeo/sync.go:373]: Note to self: the `repoName` parameter should not exist.

File: cmd/skopeo/sync.go [hunk 5]
```diff
@@ -278,21 +279,82 @@ func imagesToCopyFromRegistry(registryName string, cfg registrySyncConfig, sourc
...
+				}
+			} else {
+				// nil is equl full tags
+				if tags != nil {
+					tagList = tags.([]string)
+				}
+			}
+
+			for _, tag := range tagList {
+				source := fmt.Sprintf("%s:%s", repoName, tag)
```
Comment (mtrmac) [inline cmd/skopeo/sync.go:308]: Note to self: This should use the reference API.

**Pull request merged**